faculty/clients/api.py

```diff
@@ -0,0 +1,635 @@
+        subdomain,
+        name="",
+        description="",
+        environment_ids=[],
+        default_server_size=None,
+    ):
+        """Create a new API.
+
+        Parameters
+        ----------
+        project_id : uuid.UUID
+            The project to create the API in.
+        api_definition : CommandDefinition
+            The API's command definition: WSGI(Flask)/Plumber/Script(Custom).
+        subdomain : str
+            The subdomain where the API should run.
+        name : str
+            The API's name.
+        description : str
+            The API's description field.
+        environment_ids : List[uuid.UUID]
+            The environments to apply to the API's instances.
+        default_server_size : ServerType
+            The default server size to set
+        """
+        payload = {
+            "definition": CommandDefinitionSchema().dump(api_definition)
+        }
+        payload["subdomain"] = subdomain
+        payload["name"] = name
+        payload["description"] = description
+
+        payload["environmentIds"] = environment_ids
+
+        # TODO: turn these into schema handling as well
+        if default_server_size is None:
+            default_server_size = {
+                "instanceSizeType": "custom",
+                "instanceSize": {"milliCpus": 1000, "memoryMb": 4096},
+            }
+        payload["defaultServerSize"] = default_server_size
+
+        return self._post_raw(
+            "/project/{}/api".format(project_id), json=payload
+        )
+
+    @staticmethod
+    def flask_definition(working_directory, module, wsgi_object):
+        return WSGIDefinition(
+            api_type=APIType.WSGI,
+            working_directory=working_directory,
+            module=module,
+            wsgi_object=wsgi_object,
+        )
+
+    @staticmethod
+    def plumber_definition(working_directory, script_name):
+        return PlumberDefinition(
+            api_type=APIType.PLUMBER,
+            working_directory=working_directory,
+            script_name=script_name,
+        )
+
+    @staticmethod
+    def script_definition(working_directory, script_name):
+        return ScriptDefinition(
+            api_type=APIType.SCRIPT,
+            working_directory=working_directory,
+            script_name=script_name,
+        )
+
+    def _get_current_api_definition(self, project_id, api_id):
+        current_api = self.get(project_id, api_id)
+        current_api = _APISchema().dump(current_api)
+        return current_api
+
+    def get(self, project_id, api_id):
+        return self._get(
+            "/project/{}/api/{}".format(project_id, api_id), _APISchema()
+        )
+
+    def update(
+        self,
+        project_id,
+        api_id,
+        environment_ids=None,
+        api_definition=None,
+        server_type=None,
+        current_api=None,
+    ):
+        if not current_api:
+            current_api = self._get_current_api_definition(project_id, api_id)
```

Comment:
This kind of logic should be outside the client - the client methods are intended to map exactly on to single endpoints.